<h1><center>Laboratorio 4: El Pandas no muerde (act. II) 🐼</center></h1>

<center><strong>MDS7202: Laboratorio de Programación Científica para Ciencia de Datos</strong></center>

### Cuerpo Docente:

- Profesor: Pablo Badilla
- Auxiliar: Ignacio Meza D.
- Ayudante: Constanza Peña

### Equipo: SUPER IMPORTANTE - notebooks sin nombre no serán revisados

- Nombre de alumno 1: Sebastián Rojas

### **Link de repositorio de GitHub:** `https://github.com/sebarojasmorales/MDS7202`

# Temas a tratar

- Manejo de datos tabulares usando `pandas`. En esta segunda parte se incluye adicionalmente agregaciones, concatenaciones, merge y trabajo con strings.
- Visualizaciones interactivas de los datos con `plotly`.

### Reglas:

- Fecha de entrega: 8 de octubre (atrasos hasta el 11 de octubre, 1 punto de descuento por día)
- **Grupos de máximo 2 personas**
- **Ausentes** deberán realizar la actividad solos. 
- Cualquier duda fuera del horario de clases al foro. Mensajes al equipo docente serán respondidos por este medio.
- Prohibidas las copias. 
- Pueden usar cualquer matrial del curso que estimen conveniente.


### Objetivos principales del laboratorio

- Aplicar y aprovechar las ventajas que nos ofrece la libreria `pandas`.
- Utilizar `plotly` para obtener información gráfica del dataset.
- Aplicar el **Análisis Exploratorio de Datos** a un caso en particular.


> **Nota**: El laboratorio deberá ser desarrollado sin el uso indiscriminado de iteradores nativos de python (aka "for", "while"). La idea es que aprendan a exprimir al máximo las funciones optimizadas que nos entrega `pandas`, las cuales vale mencionar, son bastante más eficientes que los iteradores nativos sobre DataFrames.

# Importamos librerias utiles 😸

In [116]:
# Libreria Core del lab.
import numpy as np
import pandas as pd 

from IPython.display import display

#Libreria para visualizar
#!pip install --upgrade plotly
import plotly.figure_factory as ff
import plotly.express as px

# 1. Rendimiento en Estudiantes 📚


<p align="center">
  <img src="https://drive.google.com/uc?export=view&id=1hZZaPQlZe4VupFOXt76_wjuUA5NpUhrw" width="300" height="300">
</p>

Para este laboratorio deberán continuar el Análisis Exploratorio de datos sobre el conjunto ```students_grades.csv```, el cual contiene una caracterización sobre el rendimiento y otros atributos de cada alumno de la *Universidad de la Cachaña* .

# 1. Carga de Datos [0.5 Puntos]

Ya finalizado en análisis inicial, ud. y su equipo le entregaron a *Don Caguayo* (rector de la Universidad de la Cachaña) tanto los resultados del análisis como también la base de datos limpia y lista para ser almacenada. Dada la ingente cantidad de los datos, el equipo de TI de la universidad resolvió separar el dataset en dos bases de datos distintas (lo que según argumentan ellos, permitiría hacer agregaciones de forma más eficiente). 

Gracias a la excelente labor de ud. y su equipo en el análisis previo, el rector le solicita continuar el trabajo con una nueva batería de análisis. Por este motivo, la sección de TI les entrega nuevamente los datos. Sin embargo, argumentan que dada una escazes de personal, solo le entregarán *dumps* (copias) de cada base de datos y su equipo deberá unir las bases de datos. Los datos se encuentran en los siguiente archivos `.json`: ´students_grades_1.json´ y ´students_grades_2.json´. 


Por ende, ud. y su equipo deciden que la primera tarea se centrará en cargar estos datos y unirlos. No se preocupe por la limpieza ni transformar el tipo de datos de las columnas, ni tampoco transformar a notas chilenas, recuerde que anteriormente ya se encargo de este tema.

In [117]:
df_grades = pd.read_json('students_grades_1.json')
df_grades2 = pd.read_json('students_grades_2.json')
df_grades=pd.concat([df_grades,df_grades2],axis=0)

Resultado esperado:

In [118]:
df_grades

names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
470     Richard Young    male        group D                 high school   
471     Wanda Russell  female        group B                 high school   
472    Marina Zeigler  female        group C           bachelor's degree   
473     Laurie Carter  female        group B            some high school   
474      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
470      standard                    none        5.14           5.50   
471  free/reduced               completed        2.38           3.64   
472  free/reduced               completed        4.96           5.44   
473      standard               completed        4.24           4.66   
474      standard               completed        5.08           5.80   

     writing score  
0             3.76  
1             5.86  
2             4.78  
3             5.56  
4             6.22  
..             ...  
470           5.26  
471           3.16  
472           5.86  
473           4.72  
474           5.56  

[875 rows x 9 columns]

## Análisis de Las Notas v2 [1.5 Puntos]

Preocupado por la dificultad que representan los boxplot, el rector le solicita buscar alternativas de visualización la distribución de las notas. 

Para esto, muestre el [boxplot](https://plotly.com/python/box-plots/) mostrado la clase anterior, utilice un [displot](https://plotly.com/python/distplot/#distplot-with-pandas) como también un [histograma con un gráfico marginal de caja](https://plotly.com/python/histograms/#visualizing-the-distribution) como también un [histograma con el ramo como faceta de fila](https://plotly.com/python/facet-plots/#histogram-facet-grids).

Luego, responda las siguientes pregunta: 

> 1. ¿Existe una diferencia notable entre las notas?

> 2. ¿Cuál de los gráficos mostrados cree que es adecuado para mostrarle al rector? ¿Y a los padres? ¿Y a un centro de estudios educativos? ¿Por qué?. Base sus respuestas en lo visto en la clase de visualizaciones como también en lo que usted y su equipo consideren correcto.

> Hint: Para elaborar el histograma, puede que le sea de utilidad hacer un `melt` del DataFrame, dejando como variables los ramos y valores las notas. Por otra parte, visiten la documentación para generar los gráficos.


**Gráfico de Caja:**

In [119]:
px.box(df_grades,y=['math score','reading score','writing score'])

**Distplot:**

In [120]:
df_melt.columns

Index(['names', 'gender', 'race/ethnicity', 'parental level of education',
       'lunch', 'test preparation course', 'variable', 'value'],
      dtype='object')

In [121]:
# use displot figurefactory
import plotly.figure_factory as ff
fig = ff.create_distplot([df_grades["math score"],df_grades["reading score"],df_grades["writing score"]], ["math score","reading score","writing score"])
fig.show()

**Histograma con Boxplots:**

In [122]:
df_melt=df_grades.melt(id_vars=['names','gender','race/ethnicity','parental level of education','lunch','test preparation course'])

In [123]:
fig = px.histogram(df_melt, x="value",color="variable",marginal="box",nbins=30)
fig.show()

**Histograma con Faceta:**

In [124]:
fig = px.histogram(df_melt, x="value", facet_col="variable",
       category_orders={"variable": ["math score", "reading score", "writing score"]},nbins=30)
fig.show()

**Justifique:**

1.- Se puede observar que las notas tienen una distribución similar para los 3 exámenes, pero se puede notar que "reading score" tiene una mediana más alta, además de que sus notas están más concentradas (se puede ver que los peores valores de este test son más altos que para los otros dos), y que "math score" tiene la mediana más baja, y que sus puntajes más bajos son menores que para los otros dos test.

2.- Cuál gráfico elegir dependería de qué se quiere mostrar realmente, sin embargo, no mostraría el displot pues se ve confuso, los color y formas se confunden entre sí, y con ello, no se puede extraer información tan fácilmente. Por su simplicidad, al rector y a los padres les mostraría el gráfico de caja, si es que se quiere mostrar de manera simple algunos datos de la concentración de notas, ya que se puede obtener fácilmente la mediana, los cuartiles, los máximos y los mínimos. Sin embargo, si se busca mostrar datos más relacionados a la forma de la distribución y las frecuencias, mostraría los histogramas con facetas.

Para el centro de estudios educativo, mostraría el histograma con gráficos de caja, pues es más completo y captura los dos puntos mencionados anteriormente. Este gráfico podría enseñarse también al rector y a los padres si se busca dar mayor información que la presentada anteriormente, pero como el histograma mezcla los valores de los 3 exámenes, podría resultar consufo para gente poco entendida del tema.


## Análisis por Nivel Educacional Etnia de los Padres [1.5 Puntos]

El rector, basado en su experiencia, cree fuertemente que el nivel educacional y la etnia de los padres influyen en las notas que obtienen sus hijos. 
Como científicos de datos, ud. y su equipo creen que deben encontrar evidencia para confirmar o refutar la hipótesis del rector.

Para esto, deciden generar dos análisis: una tabla de resumen por una parte y gráficos de caja por otro.

### Tabla de Resumen

Para generar la tabla de resumen:

- [ ] Calcular el promedio de las notas y guardarlo en una variable `GPA` (grade point average).
- [ ] Hacer una simplificación a través de un mapeo (investigar `map()`) de la variable `parental level of education`  según la siguiente conversión: 

        some high school -> school
        some college -> school
        high school -> school
        bachelor's degree -> college
        associate's degree -> college
        master's degree -> postgraduate

  Los resultados de este mapeo deben ser guardados en la columna `simple parental level of education`.


- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener el promedio de las notas. 
- [ ] Agregar según 2 niveles: `race/ethnicity` y `simple parental level of education` para obtener un conteo de los alumnos en cada grupo y agregarlos como una nueva fila count.
- [ ] Obtener el porcentaje de alumnos con respecto al total. Los porcentajes deben ser strings que contienen la frecuencia de cada grupo con respecto al total y deben ser terminados en '%'. 


Utilizar la tabla de resultados esperados como guía para desarrollar este punto.

### 1.3.4 Información por Categorías [1 punto]

In [125]:
df_grades_tabla=df_grades.copy()
df_grades_tabla['GPA']=(df_grades_tabla['math score']+df_grades_tabla['reading score']+df_grades_tabla['writing score'])/3
df_grades_tabla['simple parental level of education']=df_grades_tabla['parental level of education'].map({'some high school':'school','some college':'school','high school':'school','bachelor\'s degree':'college','associate\'s degree':'college','master\'s degree':'postgraduate'})
df_grades_agg=df_grades_tabla.groupby(['race/ethnicity','simple parental level of education']).mean().round(2)
df_grades_agg['count']=df_grades_tabla.groupby(['race/ethnicity','simple parental level of education']).count()['names']
df_grades_agg['percentage']=(df_grades_agg['count']/len(df_grades)*100).round(2).apply(str)+' %'

df_grades_agg

math score  reading score  \
race/ethnicity simple parental level of education                              
group A        college                                   4.73           5.00   
               postgraduate                              4.69           5.23   
               school                                    4.57           4.73   
group B        college                                   5.07           5.26   
               postgraduate                              4.91           5.69   
               school                                    4.69           4.89   
group C        college                                   5.02           5.37   
               postgraduate                              4.92           5.14   
               school                                    4.76           5.02   
group D        college                                   5.11           5.25   
               postgraduate                              5.22           5.54   
               school                                    5.02           5.13   
group E        college                                   5.54           5.45   
               postgraduate                              5.54           6.03   
               school                                    5.40           5.31   

                                                   writing score   GPA  count  \
race/ethnicity simple parental level of education                               
group A        college                                      4.89  4.88     24   
               postgraduate                                 5.35  5.09      2   
               school                                       4.56  4.62     51   
group B        college                                      5.19  5.18     54   
               postgraduate                                 5.55  5.38      5   
               school                                       4.76  4.78    107   
group C        college                                      5.35  5.25    102   
               postgraduate                                 5.10  5.06     15   
               school                                       4.92  4.90    155   
group D        college                                      5.25  5.20     70   
               postgraduate                                 5.73  5.50     20   
               school                                       5.11  5.09    149   
group E        college                                      5.45  5.48     52   
               postgraduate                                 5.89  5.82      6   
               school                                       5.16  5.29     63   

                                                  percentage  
race/ethnicity simple parental level of education             
group A        college                                2.74 %  
               postgraduate                           0.23 %  
               school                                 5.83 %  
group B        college                                6.17 %  
               postgraduate                           0.57 %  
               school                                12.23 %  
group C        college                               11.66 %  
               postgraduate                           1.71 %  
               school                                17.71 %  
group D        college                                 8.0 %  
               postgraduate                           2.29 %  
               school                                17.03 %  
group E        college                                5.94 %  
               postgraduate                           0.69 %  
               school                                  7.2 %

**Resultado Esperado**

|    | race/ethnicity   | simple parental level of education   |   math score |   reading score |   writing score |   GPA |   count | percentage   |
|---:|:-----------------|:-------------------------------------|-------------:|----------------:|----------------:|------:|--------:|:-------------|
|  0 | group A          | college                              |         4.74 |            5    |            4.89 |  4.88 |      24 | 2.74 %       |
|  1 |           | postgraduate                         |         4.69 |            5.23 |            5.35 |  5.09 |       2 | 0.23 %       |
|  2 |           | school                               |         4.57 |            4.73 |            4.56 |  4.62 |      51 | 5.83 %       |
|  3 | group B          | college                              |         5.07 |            5.26 |            5.19 |  5.18 |      54 | 6.17 %       |
|  4 |           | postgraduate                         |         4.91 |            5.69 |            5.55 |  5.38 |       5 | 0.57 %       |
|  5 |           | school                               |         4.69 |            4.89 |            4.76 |  4.78 |     107 | 12.23 %      |
|  6 | group C          | college                              |         5.02 |            5.37 |            5.35 |  5.25 |     102 | 11.66 %      |
|  7 |           | postgraduate                         |         4.92 |            5.14 |            5.1  |  5.06 |      15 | 1.71 %       |
|  8 |           | school                               |         4.76 |            5.02 |            4.92 |  4.9  |     155 | 17.71 %      |
|  9 | group D          | college                              |         5.11 |            5.25 |            5.25 |  5.2  |      70 | 8.0 %        |
| 10 |           | postgraduate                         |         5.22 |            5.54 |            5.73 |  5.5  |      20 | 2.29 %       |
| 11 |           | school                               |         5.02 |            5.13 |            5.11 |  5.09 |     149 | 17.03 %      |
| 12 | group E          | college                              |         5.54 |            5.45 |            5.45 |  5.48 |      52 | 5.94 %       |
| 13 |          | postgraduate                         |         5.54 |            6.03 |            5.89 |  5.82 |       6 | 0.69 %       |
| 14 |          | school                               |         5.4  |            5.31 |            5.16 |  5.29 |      63 | 7.2 %        |

## Visualizaciones [0.5 Puntos]

Ahora, implemente un gráfico de caja en donde se muestre el GPA con respecto al nivel educacional y que la variable de color sea la etnicidad y luego comente.

In [126]:
px.box(df_grades_tabla,x='simple parental level of education',y='GPA',color='race/ethnicity')

> 1. ¿Hay alguna diferencia entre los grupos graficados tanto para el nivel educacional de los padres como también para la etnicidad?
> 2. ¿Este gráfico permite hacer facilmente un análisis conjunto de estas dos variables de forma sencilla?

**Justifique:**

1.- Si bien se pueden observar cosas en común, como que para todos los niveles educacionales de los padres el grupo E tiene concentrados sus promedios en valores más altos (se puede observar que para los 3 niveles educacionales, las medianas son más altas para el grupo E), las distribuciones de los grupos son diferentes dependiendo del nivel educacional. Por ejemplo para el grupo A, los puntajes de los alumnos con padres con nivel de postgrado están sumamente concentrados, mientras que los con nivel escolar abarcan un rango de valores muy amplios. 

Otra cosa que se puede notar, es que para todos los grupos, los valores mínimos son más bajos para los alumnos con padres con nivel educacional escolar, mientras que son más altos para los con padres con postgrado.

2.- Con este gráfico se pueden comparar fácilmente las distribuciones para las dos variables estudiadas, siendo muy útil para hacer un análisis sencillo. Sin embargo, para un análisis más complejo es necesario otro método con mayor información (como la tabla), pues el gráfico omite información que puede resultar relevante, como la cantidad de personas en cada categoría.


# 1.3.5 Combinar Dataset [1 punto]

Mientras le notificaba por videollamada los resultados de sus descubrimientos a Don Caguayo, un exaltado practicante del area de TI entra a la reunión y les informa que ha encontrado una nueva base de datos que cuenta con las notas de dos asignaturas (en escala chilena): historia y ciencias. 
Para más remate, antes de huir, el practicante les cuenta que este dataframe lamentablemente contiene nuevamente los alumnos de los registros corruptos que ud. y su equipo filtraron en el análisis anterior.
 
 
El rector (evidentemente molesto por la situación) les ruega incluir estos datos (vaciados en el archivo other_grades.csv) al estudio original(`students_grades.csv`). 


Para esto, carge el archivo `other_grades.csv` y busque la forma de unir ambos DataFrames, de tal manera que las columnas de `history score` y `science score` se anexen al final del DataFrame original. **NO LIMPIE LOS DATOS**, si no que explore los distintos tipos de merge para encontrar el mas situable para su situación (y así evitar buscar duplicados).

**To-Do**

- [ ] Cargar el `other_grades.csv`
- [ ] Unir `df_grades` con `other_grades.csv` usando outer join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando left join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando right join y explique el resultado.
- [ ] Unir `df_grades` con `other_grades.csv` usando inner join y explique el resultado.
- [ ] Defina cuál join es el que utilizará para generar el nuevo DataFrame.

> **Hint**: Puede explicar los resultados del merge a través de la cantidad de filas resultantes y los valores que estas contienen.

In [127]:
df_other_grades = pd.read_csv('other_grades.csv')

outer_merged_grades = pd.merge(left=df_grades,
                           right=df_other_grades,
                           on="names",
                           how="outer")

left_merged_grades = pd.merge(left=df_grades,
                           right=df_other_grades,
                           on="names",
                           how="left")

right_merged_grades = pd.merge(left=df_grades,
                           right=df_other_grades,
                           on="names",
                           how="right")

inner_merged_grades = pd.merge(left=df_grades,
                           right=df_other_grades,
                           on="names",
                           how="inner")

print('Outer Join:')
display(outer_merged_grades)
print('Left Join:')
display(left_merged_grades)
print('Right Join:')
display(right_merged_grades)
print('Inner Join:')
display(inner_merged_grades)

Outer Join:


names  gender race/ethnicity parental level of education  \
0        Rita Courtney  female        group B            some high school   
1     Charles Linstrom    male        group A           bachelor's degree   
2          Brian Young    male        group C            some high school   
3       Howard Jimenez    male        group E            some high school   
4         Wayne Wilson    male        group B            some high school   
..                 ...     ...            ...                         ...   
995       Leona Markow     NaN            NaN                         NaN   
996  Maxine Gulbranson     NaN            NaN                         NaN   
997        Jeane Budds     NaN            NaN                         NaN   
998        Nancy Ramos     NaN            NaN                         NaN   
999     Andrea Mcdavid     NaN            NaN                         NaN   

        lunch test preparation course  math score  reading score  \
0    standard                    none        3.22           3.76   
1    standard               completed        5.80           5.68   
2    standard                    none        5.38           4.96   
3    standard               completed        5.86           5.50   
4    standard               completed        6.64           6.16   
..        ...                     ...         ...            ...   
995       NaN                     NaN         NaN            NaN   
996       NaN                     NaN         NaN            NaN   
997       NaN                     NaN         NaN            NaN   
998       NaN                     NaN         NaN            NaN   
999       NaN                     NaN         NaN            NaN   

     writing score  science score  history score  
0             3.76             25            3.8  
1             5.86             56            5.2  
2             4.78             58            3.8  
3             5.56             72            5.0  
4             6.22             94            4.6  
..             ...            ...            ...  
995            NaN             39            3.8  
996            NaN             38            3.9  
997            NaN             43            3.9  
998            NaN             54            6.2  
999            NaN             40            4.2  

[1000 rows x 11 columns]

Left Join:


names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
870      standard                    none        5.14           5.50   
871  free/reduced               completed        2.38           3.64   
872  free/reduced               completed        4.96           5.44   
873      standard               completed        4.24           4.66   
874      standard               completed        5.08           5.80   

     writing score  science score  history score  
0             3.76             25            3.8  
1             5.86             56            5.2  
2             4.78             58            3.8  
3             5.56             72            5.0  
4             6.22             94            4.6  
..             ...            ...            ...  
870           5.26             62            4.2  
871           3.16             16            3.6  
872           5.86             59            4.1  
873           4.72             43            4.3  
874           5.56             61            5.3  

[875 rows x 11 columns]

Right Join:


names  gender race/ethnicity parental level of education  \
0           Pam Otoole  female        group B           bachelor's degree   
1          Diane Olson  female        group C                some college   
2       Sarah Flannery  female        group B             master's degree   
3     Stanley Chappell    male        group A          associate's degree   
4        Arnold Dodson    male        group C                some college   
..                 ...     ...            ...                         ...   
995    Melanie Acevedo  female        group E             master's degree   
996   Christopher Wall    male        group C                 high school   
997  Margaret Spannaus  female        group C                 high school   
998     Vickie Sellers  female        group D                some college   
999      Kari Mitchell  female        group D                some college   

            lunch test preparation course  math score  reading score  \
0        standard                    none        5.32           5.32   
1        standard               completed        5.14           6.40   
2        standard                    none        6.40           6.70   
3    free/reduced                    none        3.82           4.42   
4        standard                    none        5.56           5.68   
..            ...                     ...         ...            ...   
995      standard               completed        6.28           6.94   
996  free/reduced                    none        4.72           4.30   
997  free/reduced               completed        4.54           5.26   
998      standard               completed        5.08           5.68   
999  free/reduced                    none        5.62           6.16   

     writing score  science score  history score  
0             5.44             57            4.0  
1             6.28             48            4.8  
2             6.58             72            6.1  
3             3.64             47            3.7  
4             5.50             60            5.2  
..             ...            ...            ...  
995           6.70             79            6.3  
996           4.30             62            4.0  
997           4.90             47            4.0  
998           5.62             47            5.7  
999           6.16             53            6.2  

[1000 rows x 11 columns]

Inner Join:


names  gender race/ethnicity parental level of education  \
0       Rita Courtney  female        group B            some high school   
1    Charles Linstrom    male        group A           bachelor's degree   
2         Brian Young    male        group C            some high school   
3      Howard Jimenez    male        group E            some high school   
4        Wayne Wilson    male        group B            some high school   
..                ...     ...            ...                         ...   
870     Richard Young    male        group D                 high school   
871     Wanda Russell  female        group B                 high school   
872    Marina Zeigler  female        group C           bachelor's degree   
873     Laurie Carter  female        group B            some high school   
874      Amanda Perez  female        group A                 high school   

            lunch test preparation course  math score  reading score  \
0        standard                    none        3.22           3.76   
1        standard               completed        5.80           5.68   
2        standard                    none        5.38           4.96   
3        standard               completed        5.86           5.50   
4        standard               completed        6.64           6.16   
..            ...                     ...         ...            ...   
870      standard                    none        5.14           5.50   
871  free/reduced               completed        2.38           3.64   
872  free/reduced               completed        4.96           5.44   
873      standard               completed        4.24           4.66   
874      standard               completed        5.08           5.80   

     writing score  science score  history score  
0             3.76             25            3.8  
1             5.86             56            5.2  
2             4.78             58            3.8  
3             5.56             72            5.0  
4             6.22             94            4.6  
..             ...            ...            ...  
870           5.26             62            4.2  
871           3.16             16            3.6  
872           5.86             59            4.1  
873           4.72             43            4.3  
874           5.56             61            5.3  

[875 rows x 11 columns]

> **Justificación:**


El *outer join* deja los datos de ambos datasets, por lo que los nombres "erróneos" (que no están en el dataset original limpiado) se agregan igual, incluyendo solamente las notas del segundo dataset (puntaje de historia y ciencias), y aumentando así el número de filas a 1000.

El *inner join* deja a las personas que estén en ambos datasets, por lo que en este caso, como el nuevo dataset encontrado contiene a todas las personas del dataset original (limpiado) y otras erróneas, se borran las erróneas y se agregan las notas de los 5 ramos para las personas del dataset original (manteniéndose las 875).

El *left join* (aplicando el dataset original a la izquierda) dejará a todas las personas del dataset original junto a las 3 notas originales, mientras que agregará las notas de historia y ciencias para las que, además de estar en el dataset original, están en el nuevo dataset encontrado (a las que no, les rellena estas notas con NaN). Como ya dijimos, todas las personas del dataset original están en el nuevo, por lo que el *left join* funciona como el *inner join*.

El *right join* (aplicando el dataset original a la izquierda) dejará a todas las personas del dataset nuevo, junto a sus notas de historia y ciencias, y agregará las notas de matemáticas, lectura y escritura para las personas del dataset nuevo que también están en el original (a las que no, les rellena estas notas con NaN). Como ya dijimos, todas las personas del dataset original están en el nuevo, por lo todas ellas quedarán con sus 5 notas, y el resto quedará con las de historia y ciencias, es decir, funciona igual al *outer join*.


Dado lo anterior, para este caso serviría tanto el *left join* (si se ocupa el dataset original a la izquierda, y si se ocupa a la derecha, serviría el *right join*) como el *inner join* para generar el nuevo DataFrame. Eso sí, para un caso un poco más general, dependería de si se quieren mantener solamente los que tengan notas de los 5 ramos (en este caso habría que ocupar *inner join*) o si se quiere mantener de todas maneras a las personas que estaban en el dataset original limpiado a pesar de que en el nuevo dataset encontrado (el de las notas de historia y ciencias) no estén sus notas (en este caso habría que ocupar *left join*, tomando el dataset original a la izquierda, o *right join* con el dataset original a la derecha, pero cabe notar que en este problema esto no pasa, pues el nuevo dataset encontrado contiene a todos las personas del dataset original limpiado).

## 1.4 Más visualizaciones [Bonus - 0.5 puntos por cada Gráfico, máximo 1 punto]


**Desde aquí parte el bonus para el lab.**


<p align="center">
  <img src="https://media4.giphy.com/media/xT5LMWNOjGqJzUfyve/giphy.gif?cid=ecf05e47phkhvg6vadqvcoobvxwj9bzxxk38jjvkcc6eak6t&rid=giphy.gif&ct=g">
</p>




Genere dos visualizaciones extras que encuentre interesantes (y no triviales) con estos datos y explique sus resultados. Agrupe los atributos que estime convenientes.


**To-Do:**
- [ ] Generar dos nuevas visualizaciones con los datos y explicar que están representando.


> **NOTA: No utilice historia ni ciencias, son notas generadas aleatoriamente.**

# Conclusión
Eso ha sido todo para el lab de hoy, recuerden que el laboratorio tiene un plazo de entrega de una semana y que **los días de atraso no se pueden utilizar para entregas de lab** solo para tareas. Cualquier duda del laboratorio, no duden en contactarnos por mail o U-cursos.

<p align="center">
  <img src="https://media1.tenor.com/images/2f7df460196625edc45b8380d158e5ee/tenor.gif?itemid=16669403">
</p>

<br>
<center>
<img src="https://i.kym-cdn.com/photos/images/original/001/194/195/b18.png" width=100 height=50 />
</center>
<br>